k-distribution/k-tutorial/1_basic/03_parsing/README.md

```diff
@@ -1 +1,362 @@
+        Location(1,1,1,23)
+```
+
+This error is saying that `kast` was unable to parse this program because it is
+ambiguous. K's just-in-time parser is a GLL parser, which means it can handle
+the full generality of context-free grammars, including those grammars which
+are ambiguous. An ambiguous grammar is one where the same string can be parsed
+as multiple distinct ASTs. In this example, it can't decide whether it should
+be parsed as `(true && false) || false` or as `true && (false || false)`. As a
+result, it reports the error to the user.
+
+## Brackets
+
+Currently there is no way of resolving this ambiguity, making it impossible
+to write complex expressions in this language. This is obviously a problem.
+The standard solution in most programming languages to this problem is to
+use parentheses to indicate the appropriate grouping. K generalizes this notion
+into a type of production called a **bracket**. A bracket production in K
+is any production with the `bracket` attribute. It is required that such a
+production only have a single non-terminal, and the sort of the production
+must equal the sort of that non-terminal. However, K does not otherwise
+impose restrictions on the grammar the user provides for a bracket. With that
+being said, the most common type of bracket is one in which a non-terminal
+is surrounded by terminals representing some type of bracket such as
+`()`, `[]`, `{}`, `<>`, etc. For example, we can define the most common
+type of bracket, the type used by the vast majority of programming languages,
+quite simply.
+
+Consider the following modified definition, which we will save to
+`lesson-03-d.k`:
+
+```k
+module LESSON-03-D
+
+  syntax Boolean ::= "true" | "false"
+                   | "(" Boolean ")" [bracket]
+                   | "!" Boolean [function]
+                   | Boolean "&&" Boolean [function]
+                   | Boolean "^" Boolean [function]
+                   | Boolean "||" Boolean [function]
+
+endmodule
+```
+
+In this definition, if the user does not explicitly define parentheses, the
+grammar remains ambiguous and K's just-in-time parser will report an error.
+However, you are now able to parse more complex programs by means of explicitly
+grouping subterms with the bracket we have just defined.
+
+Consider `and-or-left.bool`:
+
+```
+(true && false) || false
+```
+
+Now consider `and-or-right.bool`:
+
+```
+true && (false || false)
+```
+
+If you parse these programs with `kast`, you will once again get a single
+unique AST with no error. If you look, you might notice that the bracket itself
+does not appear in the AST. In fact, this is a property unique to brackets:
+productions with the bracket attribute are not represented in the parsed AST
+of a term, and the child of the bracket is folded immediately into the parent
+term. This is the reason for the requirement that a bracket production have
+a single non-terminal of the same sort as the production itself.
```

Comment:
```suggestion
of a term, and the child of the bracket is folded immediately into the parent
term. This is the reason for the requirement that a bracket production have
a single non-terminal of the same sort as the production itself.

Confirm using `kast` that indeed, the bracket production defined above does
not appear in the resulting output kore.
```

Reply:
This is redundant; we already talk about using kast and noticing that the bracket doesn't appear in the resulting AST.

Reply:
We talk about it, but this is about giving the user an exercise (so you could put the `### Exercise` block here). I think many people learn better if they are given little tasks to confirm their understanding, so that they have to figure out "what do I need to type in the terminal to confirm that the brackets do not show up?" and "how do I read the output and make sure that the bracket does not show up?"

Reply:
Ah I see you gave exactly this exercise below. Sounds good!